prober/derp.go

```diff
@@ -577,6 +580,196 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
 	return nil
 }
 
+// derpProbeBandwidthMu ensures that derpProbeBandwidthTUN doesn't run concurrently.
+// This is necessary to avoid conflicts trying to create the TUN device.
+var derpProbeBandwidthMu sync.Mutex
+
+// derpProbeBandwidthTUN takes two DERP clients (fromc and toc) connected to two
+// DERP servers (from and to) and sends a test payload of a given size from one
+// to another over TUN.
+func derpProbeBandwidthTUN(ctx context.Context, transferTime *expvar.Float, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, size int64) error {
+	derpProbeBandwidthMu.Lock()
+	defer derpProbeBandwidthMu.Unlock()
+
+	// Temporarily set up a TUN device with which to simulate a real client TCP connection
+	// tunneling over DERP.
+	dev, _, err := tstun.New(log.Printf, "utun9")
+	if err != nil {
+		return fmt.Errorf("failed to create TUN device: %w", err)
+	}
+	defer dev.Close()
+	mtu, err := dev.MTU()
+	if err != nil {
+		return fmt.Errorf("failed to get TUN MTU: %w", err)
+	}
+
+	r, err := router.New(log.Printf, dev, nil, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create router: %w", err)
+	}
+	defer r.Close()
+
+	err = r.Set(&router.Config{
+		LocalAddrs: []netip.Prefix{netip.MustParsePrefix("172.16.0.1/30")},
+		Routes:     []netip.Prefix{netip.MustParsePrefix("172.16.0.1/30")},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to set routes: %w", err)
+	}
+
+	// Read packets from TUN device
+	tunReadC := make(chan error, 1)
+	go func() {
+		// There's not a lot of thought in the below number. 5 doesn't use a ton of memory,
+		// and a bigger number like 20 didn't seem to increase throughput.
+		numBufs := 5
+		bufs := make([][]byte, 0, numBufs)
+		sizes := make([]int, numBufs)
+		for range numBufs {
+			bufs = append(bufs, make([]byte, mtu+tstun.PacketStartOffset))
+		}
+
+		for {
+			n, err := dev.Read(bufs, sizes, tstun.PacketStartOffset)
+			if err != nil {
+				return
+			}
+
+			for i := range n {
+				buf := bufs[i][:sizes[i]+tstun.PacketStartOffset]
+				pkt := buf[tstun.PacketStartOffset:]
+				src, dst := make([]byte, 4), make([]byte, 4)
+				srcPort, dstPort := make([]byte, 2), make([]byte, 2)
```

Comment:
These could be reused across read ops, though toolchain might be doing that for us. We can move them outside the read loop.